lib/Sema/TypeCheckSwitchStmt.cpp

```diff
@@ -696,19 +698,38 @@ namespace {
           buffer << (getBoolValue() ? "true" : "false");
           break;
         case SpaceKind::Constructor: {
-          if (!Head.empty()) {
+          if (!Head.getBaseIdentifier().empty()) {
             buffer << ".";
-            buffer << Head.str();
+            buffer << Head.getBaseIdentifier().str();
           }
 
           if (Spaces.empty()) {
             return;
           }
 
+          llvm::SmallVector<std::pair<Identifier, Space>, 4> labelSpaces;
+          auto args = Head.getArgumentNames().begin();
+          auto arg_end = Head.getArgumentNames().end();
+          for (auto param : Spaces) {
+            if (args != arg_end) {
+              labelSpaces.push_back(std::pair<Identifier, Space>(*args, param));
+              args++;
+            } else
+              labelSpaces.push_back(
```

Comment:
Could you assert that you always have the same number of argument names in the base identifier as there are subspaces?  If that much is true, then you can swap the interleave for the variant of `std::transform` that takes two input iterators and keep track of placing the commas yourself.  Honestly, if it's all the same we can just take this code and you can leave a FIXME behind to come clean this up.  I'm not too concerned about printing being efficient since this will only come up when you're diagnosing invalid code.

Reply:
Do you mean a full assertion that stops compilation? Or just a check that runs a different path if that is the case?

Reply:
Yeah, just an `assert`.

Reply:
Is that really what we want to happen though? Because there could be a case where there are no argument names but there are spaces, so we wouldn't want it to stop in that case.

Reply:
I am not sure just having an assertion allows all cases we would want.

For instance, if we check the number of argument names against spaces in the following example:
```enum Threepeat {
  case a, b, c
}

func test3(x: Threepeat, y: Threepeat) {
  switch (x, y) { // expected-error {{switch must be exhaustive}}
  // expected-note@-1 {{add missing case: '(.a, .c)'}}
  case (.a, .a):
    ()
  case (.b, _):
    ()
  case (.c, _):
    ()
  case (_, .b):
    ()
  }
}```
with the assertion, it fails where it should not.